### PR TITLE
Rest of the UI

### LINKED ISSUE #35: Implement mouse and rest of UI
- [ ] Left click on ship, right click on map = move command - get id of clicked ship, get the vector from ship to second click and limit to 1 unit magnitude.
- [ ] Left click on ship, left click on ship = attack command -  get id of first ship, get id of second ship, have first attack second
- [ ] Extra state - IORef Maybe Entity - currently selected ship - remember to use locks
- [ ] Up arrow key = send move list - send list of moves to server, read new board from server.

### PR DESCRIPTION
**Pull request by JohannCooper**
Closes #35

### CODE REVIEW

File: src/client/Interaction.hs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+import Theme
+import Util
+
+
+windowToWorld :: Vec2 -> Vec2
+windowToWorld (x, y) = worldPos
+  where
+    canvasX = (2 * x - fromIntegral windowWidth) / fromIntegral windowWidth
+    canvasY = (-2 * y + fromIntegral windowHeight) / fromIntegral windowHeight
+    worldPos = (canvasX * mapBorder, canvasY * mapBorder)
```
Comment (JustinHuPrime) [inline src/client/Interaction.hs:40]: Shouldn't you do a beta-reduction on worldPos? i..e.
```suggestion
windowToWorld (x, y) = canvasX * mapBorder, canvasY * mapBorder
  where
    canvasX = (2 * x - fromIntegral windowWidth) / fromIntegral windowWidth
    canvasY = (-2 * y + fromIntegral windowHeight) / fromIntegral windowHeight
```

File: src/client/Interaction.hs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+isMouseOverShip :: Position -> Entity -> Bool
+isMouseOverShip (Position x y) (Ship _ shipPos _ _ _ _ _ _ _ _ _ _) = distance shipPos mouseWorldPos <= hitboxRadius
```
Comment (JustinHuPrime) [inline src/client/Interaction.hs:43]: You can convert the ship constructor pattern match into a record pattern match

File: src/client/Interaction.hs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  where
+    mouseWorldPos = windowToWorld (fromIntegral x, fromIntegral y)
+    hitboxRadius = 0.5
```
Comment (JustinHuPrime) [inline src/client/Interaction.hs:46]: This should be a constant somewhere

File: src/client/Interaction.hs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+isMouseOverShip _ _ = False
+
+getClickedShip :: Key -> KeyState -> Position -> Board -> Maybe Entity
+getClickedShip (MouseButton LeftButton) Up mousePos board = entity
+  where
+    entity = if null clickedShips then Nothing else Just (head clickedShips)
+    clickedShips = filter (isMouseOverShip mousePos) board
```
Comment (JustinHuPrime) [inline src/client/Interaction.hs:53]: Again, shouldn't a beta-reduction happen for entity?
```suggestion
getClickedShip (MouseButton LeftButton) Up mousePos board = if null clickedShips then Nothing else Just (head clickedShips)
  where
    clickedShips = filter (isMouseOverShip mousePos) board
```

File: src/client/Interaction.hs [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+getClickedShip _ _ _ _ = Nothing
+
+getClickedMapPos :: Key -> KeyState -> Position -> Maybe Vec2
+getClickedMapPos (MouseButton RightButton) Up (Position x y) = Just (windowToWorld (fromIntegral x, fromIntegral y))
+getClickedMapPos _ _ _ = Nothing
+
+handleMoves :: Board -> IORef [Move] -> Lock.Lock -> IORef (Maybe Entity) -> Lock.Lock -> Key -> KeyState -> Position -> IO ()
+handleMoves board moveList moveListLock selectedEntity selectedEntityLock key keyState mousePos = do
+  let clickedShip = getClickedShip key keyState mousePos board
+  let clickedMapPos = getClickedMapPos key keyState mousePos
```
Comment (JustinHuPrime) [inline src/client/Interaction.hs:63]: Shouldn't this be `let ... in`?

File: src/client/Interaction.hs [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+  Lock.acquire moveListLock
+  Lock.acquire selectedEntityLock
+
+  moveList' <- readIORef moveList
+  selectedEntity' <- readIORef selectedEntity
+
+  -- -- select new player ship
```
Comment (JustinHuPrime) [inline src/client/Interaction.hs:71]:
```suggestion
  -- select new player ship
```

File: src/client/Interaction.hs [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    let (sX, sY) = Board.position (fromJust selectedEntity')
+    let (mX, mY) = fromJust clickedMapPos
+    let dv = vecUnit (mX - sX, mY - sY)
+
+    let newMove = Thrust idNum dv
+    let isNotDuplicate (Thrust id _) = id /= idNum 
+        isNotDuplicate Attack {} = True
+    let filteredMoveList = filter isNotDuplicate moveList'
+
+    writeIORef moveList (newMove:filteredMoveList)
```
Comment (JustinHuPrime) [inline src/client/Interaction.hs:87]: Shouldn't this be beta reduced? i.e. substitute in all of those lets?

File: src/client/Interaction.hs [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    writeIORef selectedEntity Nothing
+    print "thrust command queued"
```
Comment (JustinHuPrime) [inline src/client/Interaction.hs:89]:
```suggestion
```

File: src/client/Interaction.hs [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+      let newMove = Attack attacker target
+      let isNotDuplicate (Attack id _) = id /= attacker 
+          isNotDuplicate Thrust {} = True
+      let filteredMoveList = filter isNotDuplicate moveList'
+
+      Control.Monad.when (attacker /= target) $ do
+        writeIORef moveList (newMove:filteredMoveList)
+        writeIORef selectedEntity Nothing
+        print "attack command queued"
```
Comment (JustinHuPrime) [inline src/client/Interaction.hs:104]:
```suggestion
```

File: src/client/Interaction.hs [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+      let attacker = Board.idNum (fromJust selectedEntity')
+      let target = Board.idNum (fromJust clickedShip)
+
+      let newMove = Attack attacker target
+      let isNotDuplicate (Attack id _) = id /= attacker 
+          isNotDuplicate Thrust {} = True
+      let filteredMoveList = filter isNotDuplicate moveList'
+
+      Control.Monad.when (attacker /= target) $ do
+        writeIORef moveList (newMove:filteredMoveList)
```
Comment (JustinHuPrime) [inline src/client/Interaction.hs:102]: As above, shouldn't this be beta reduced?

File: src/client/Interaction.hs [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    let isNotDuplicate (Thrust id _) = id /= idNum 
+        isNotDuplicate Attack {} = True
+    let filteredMoveList = filter isNotDuplicate moveList'
+
+    writeIORef moveList (newMove:filteredMoveList)
+    writeIORef selectedEntity Nothing
+    print "thrust command queued"
+
+  -- -- attack enemy ship
+  Control.Monad.when (isJust selectedEntity' && isJust clickedShip) $ do
```
Comment (JustinHuPrime) [inline src/client/Interaction.hs:92]: You should probably use case statements to match on selectedEntity' and clickedShip

File: src/client/Interaction.hs [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+
+  moveList' <- readIORef moveList
+  selectedEntity' <- readIORef selectedEntity
+
+  -- -- select new player ship
+  Control.Monad.when (isJust clickedShip) $
+      writeIORef selectedEntity clickedShip
+
+  -- thrust toward board position
+  Control.Monad.when (isJust selectedEntity' && isJust clickedMapPos) $ do
```
Comment (JustinHuPrime) [inline src/client/Interaction.hs:76]: You should probably use a case statement to decompose selectedEntity' and clickedMapPos

File: src/client/Interaction.hs [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+  Lock.release selectedEntityLock
+  Lock.release moveListLock
+
+sendMovesToServer :: Socket -> IORef Board -> Lock.Lock -> IORef [Move] -> Lock.Lock -> Key -> KeyState -> IO ()
+sendMovesToServer socket board boardLock moveList moveListLock (SpecialKey KeyUp) Up = do
+  Lock.acquire moveListLock
+  moveList' <- readIORef moveList
+  sendMoves socket moveList'
+  writeIORef moveList []
+  print "sent"
```
Comment (JustinHuPrime) [inline src/client/Interaction.hs:115]:
```suggestion
```

File: src/client/Interaction.hs [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+  updateBoardState socket board boardLock
```
Comment (JustinHuPrime) [inline src/client/Interaction.hs:116]: This should probably be beta-substituted
```suggestion
  Lock.acquire boardLock
  writeIORef board <$> readBoard socket
  Lock.release boardLock
```

File: src/client/Interaction.hs [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+  print "board updated"
+  Lock.release moveListLock
+sendMovesToServer _ _ _ _ _ _ _ = return ()
+
+updateBoardState :: Socket -> IORef Board -> Lock.Lock -> IO ()
+updateBoardState socket board boardLock = do
+  Lock.acquire boardLock
+  board' <- readBoard socket
+  writeIORef board board'
```
Comment (JustinHuPrime) [inline src/client/Interaction.hs:125]: Can't these two be collapsed with the monadic composition operator?
```suggestion
  writeIORef board <$> readBoard socket
```

File: src/client/Renderer.hs [hunk 16]
```diff
@@ -18,41 +18,45 @@ with N-Planetary. If not, see <https://www.gnu.org/licenses/>.
...
-
-renderCircle :: Vec2 -> Double -> Double -> Int -> IO ()
-renderCircle position radius layer vertices = renderFan (position : circle position radius vertices) layer
-
-renderBoard :: Board -> IO ()
-renderBoard = mapM_ renderEntity
+    angles    = map ((*arc) . fromIntegral) [0..numVertices]
+    points    = center : map toPoint angles
+    vertices  = mapM_ (\(x, y) -> vertex (Vertex3 x y layer)) points
+    color'    = color $ Color3 r g b
```
Comment (JustinHuPrime) [inline src/client/Renderer.hs:38]: What's `color`? I guess it's a function, right? Which library is it from? (I'm a bit uncertain about the libraries used - your code is fine)

File: src/client/Renderer.hs [hunk 17]
```diff
@@ -18,41 +18,45 @@ with N-Planetary. If not, see <https://www.gnu.org/licenses/>.
...
-renderFan :: [(Double, Double)] -> Double -> IO ()
-renderFan points layer = renderPrimitive TriangleFan $ mapM_ (\(x, y) -> vertex (Vertex3 x y layer)) points
-
-renderCircle :: Vec2 -> Double -> Double -> Int -> IO ()
-renderCircle position radius layer vertices = renderFan (position : circle position radius vertices) layer
-
-renderBoard :: Board -> IO ()
-renderBoard = mapM_ renderEntity
+    angles    = map ((*arc) . fromIntegral) [0..numVertices]
+    points    = center : map toPoint angles
```
Comment (JustinHuPrime) [inline src/client/Renderer.hs:36]: There's definitely some beta reduction to be done here.

File: src/client/Renderer.hs [hunk 18]
```diff
@@ -18,41 +18,45 @@ with N-Planetary. If not, see <https://www.gnu.org/licenses/>.
...
 
 -- TODO: add colors from theme
 renderEntity :: Entity -> IO ()
-renderEntity (AstroObj _ (x, y) _ _ radius) = renderCircle position radius' layer 32
+renderEntity (AstroObj _ (x, y) _ _ r) = renderCircle center radius 32 color layer
   where
-    position = (x / fromIntegral windowWidth, y / fromIntegral windowHeight)
-    radius' = radius / fromIntegral windowWidth
+    center = worldToCanvas (x, y)
+    radius = fst (worldToCanvas (r, 0))
```
Comment (JustinHuPrime) [inline src/client/Renderer.hs:45]: Clever! But shouldn't that sort of thing have its own function?

**Review by JustinHuPrime — CHANGES_REQUESTED** [review]

File: src/client/Interaction.hs (hunk 1 above)
Comment (JohannCooper) [reply src/client/Interaction.hs:40]: I think both are fine but I know this works as is. The system will be refactored anyways with transformation matrices.

File: src/client/Interaction.hs (hunk 2 above)
Comment (JohannCooper) [reply src/client/Interaction.hs:43]: Oh didn't know that existed. I will leave it as is until refactors.

File: src/client/Interaction.hs (hunk 4 above)
Comment (JohannCooper) [reply src/client/Interaction.hs:53]: See above comment

File: src/client/Interaction.hs (hunk 5 above)
Comment (JohannCooper) [reply src/client/Interaction.hs:63]: yes but I want to use these throughout the entire function

File: src/client/Interaction.hs (hunk 15 above)
Comment (JohannCooper) [reply src/client/Interaction.hs:125]: Same as above

File: src/client/Interaction.hs (hunk 14 above)
Comment (JohannCooper) [reply src/client/Interaction.hs:116]: Same as above

File: src/client/Renderer.hs (hunk 18 above)
Comment (JohannCooper) [reply src/client/Renderer.hs:45]: Yep but will be overhauled with transformation matrices

File: src/client/Renderer.hs (hunk 17 above)
Comment (JohannCooper) [reply src/client/Renderer.hs:36]: See above

**Review by JustinHuPrime — APPROVED** [review]
Approved, but noted that this generates technical debt.

**Review by wyndro — APPROVED** [review]

**Pull request merged**